projects/finder-frontend/docker-compose.yml

```diff
@@ -56,8 +51,24 @@ services:
       PLEK_SERVICE_STATIC_URI: assets.publishing.service.gov.uk
       VIRTUAL_HOST: finder-frontend.dev.gov.uk
       BINDING: 0.0.0.0
+
+  finder-frontend-app-account:
+    <<: *finder-frontend-app
+    depends_on:
+      - nginx-proxy
+      - email-alert-api-app
+      - govuk-account-manager-prototype-app
+    environment:
+      PLEK_SERVICE_SEARCH_URI: https://www.gov.uk/api
+      PLEK_SERVICE_WHITEHALL_FRONTEND_URI: https://www.gov.uk
+      GOVUK_WEBSITE_ROOT: https://www.gov.uk
+      PLEK_SERVICE_CONTENT_STORE_URI: https://www.gov.uk/api
```

Comment:
These live services seem a bit strange to me, it seems like this has become half live? I'm guessing this is desired for convenience but it would be good to be clear that what in an account stack is live or not.

It'd feel more logically consistent if these weren't live dependencies though as that'd keep it fully distinct from the live service and feel more like an extension of the app stack.

Reply:
ah true, that's just a mistake on my part - will remove those.
:thinking: given this inherits from app will it inherit:
```
      ASSET_HOST: finder-frontend.dev.gov.uk
      VIRTUAL_HOST: finder-frontend.dev.gov.uk
      BINDING: 0.0.0.0
      MEMCACHE_SERVERS: memcached
```

or do we have to declare these each time?
I see the `app` stack and `app-live` stack already have `BINDING` duplicated, do we know if that is a requirement?

Reply:
Thanks @kevindew [amended here](https://github.com/alphagov/govuk-docker/pull/431/commits/c36f5b3d5ebc1303aa2912c515e7e060426ce311)

Reply:
No it doesn't inherit any of the environment variables you have to declare them each time. It'll only inherit if we use a yaml merge within an object.

Do you need the other dependencies of the app: https://github.com/alphagov/govuk-docker/blob/c36f5b3d5ebc1303aa2912c515e7e060426ce311/projects/finder-frontend/docker-compose.yml#L26-L32 as they're not inherited either?

There is a part of me that wonders if it'd be simpler having accounts as just part of the app stack - do we anticipate that to be massively inconveniencing? I'm just wondering if we're prematurely worried and creating a snowflake.

Reply:
oh! I thought those were inherited too.
so for the checker I should need:
- router-app
- static-app
(:thinking: I did test this so I wonder how I was getting to any pages without a router...)

I suspect I can run all elements of the checker without:
- search-api-app
- content-store-app

and will need to check on
- memcached

This presents an interesting dilemma though, with different dependencies `app-account` would not be an enhancement to the `app` stack... instead it would accept that instead of the entire app you're only interested in the accounts bit... so anyone running `app-accounts` who goes over to a proper finder or search, would find that not working... is that ok...? :thinking:

Reply:
> There is a part of me that wonders if it'd be simpler having accounts as just part of the app stack - do we anticipate that to be massively inconveniencing? I'm just wondering if we're prematurely worried and creating a snowflake.

Yeah I'm open to this idea, and also bit concerned my snow-flake-y-ness too. The pros as I see them are:

- I strongly suspect our trial will continue to other applications at some point soon (new or existing tbc)
- This should result in less disruption for those who are also working in these apps (e.g. Brexit Transition folks working on the checker will not neet to get accounts set up if they're still running the traditional `app` stack)... so it's a concious opt-in to accounts which seems nice to me...
- For apps that have a `live` stack i think it clearly sets these out as something that has no overlap... so you can't have an `app-live` with working accounts (I'm assuming it would be confusing to get real data from the content store... but be speaking to a local development account)

Does that sway you at all? Or do you think we should put these back in `app` and then deal with this when there's a more presing need?

Reply:
Thanks Huw. Accounts does present an interesting dilemma for the live stack where before it represented everything being live bar the application, whereas it will now mean almost everything live 🤔  unless we're able to work out how apps can integrate without accounts without needing secrets for accounts (is that a goal? would be ace)

I'm not massively opinionated either way so can be persuaded that an app-account stack is the way to go. I am leaning more towards just putting accounts in the general app stack though as it feels like this is more conventional with existing practices and opens up less ambiguities. It feels like something that should be considered further if evidence suggests it does inconvenience people, and if that's the case it might be a question of "is accounts unnecessarily hard to run?"

Reply:
I suspect dev-live will be a consistent issue - the applications that speak to accounts need to be "registered" with the account manager and given the right shared secret keys to prove they're authorised to do the things they're requesting - and to prevent anyone and their cat from firing off production requests.

So unless we had something where the "live" stack could speak to a remote server (staging?) and we were happy for those staging secrets to leak into dev (... which I can't get a really good use case for right now...) then I suspect there'll continue to be a rift between live and development setups.

This is probably a larger thing for us to consider, perhaps at the upcoming tech fortnightly? I guess many of our tools have been predicated on us having unauthenticated APIs in production and transactional services. Perhaps Accounts is the question of what do we do when that assumption no longer holds... (and how likely is that at this point?)

Interesting interesting...
I will catch up with my team and get their review here, and if you are flexible, then I'll do one of two things:
- If my team likes `app-account` I'll merge this and then keep an eye on it, at the end of our trial I'm happy to revert this PR and then address the larger question of accounts vs our dev tools when we've committed to services that aren't in alpha?
- If my team isn't excited about `app-account` then I'll follow your suggestion, move this back into `app` and put a shout out to anyone else working in `finder-frontend` that if they're pulling master they'll need our two repos too.

I think those questions about how easy/difficutly accounts might be to setup are interesting ones, throw them our way during out acconts tech fortnightly update?

Reply:
Great stuff! main ask if you go with an app-accounts stack is that you're clear about what should and shouldn't go in it to try avoid confusion over some apps being live APIs and others not.

Yeah with regards to how it'll work in live stacks, we have dealt with similar with signon by apps using a gem to interact and the mock environment for development. Maybe some inspiration could be taken from how those problems were solved if it's applicable?

Reply:
> Accounts does present an interesting dilemma for the live stack where before it represented everything being live bar the application, whereas it will now mean almost everything live 🤔 unless we're able to work out how apps can integrate without accounts without needing secrets for accounts (is that a goal? would be ace)

There's perhaps some ideas we can take from gds-sso here, where when running in development mode the user gets signed into the first account.  Maybe we can make some secrets optional:

- `GOVUK_ACCOUNT_JWT_KEY_PEM` and `GOVUK_ACCOUNT_JWT_KEY_UUID` - I think we could make signing the JWT optional in dev mode.  Currently we only allow one cipher, but we could allow that + (if dev) no cipher.
- `GOVUK_ACCOUNT_OAUTH_CLIENT_ID` and `GOVUK_ACCOUNT_OAUTH_CLIENT_SECRET` - these are trickier because the client ID identifies which app the user is coming from, but maybe we could make these more consistent and able to have a sensible default value in dev?

How exactly GOV.UK apps are going to integrate with accounts when it gets rolled out to more things is still fairly unknown.  We've currently got special-case code in finder-frontend and collections, but if we make a new gem, or move the common code to govuk_app_config, then we could put this dev secrets management logic there.  I think we should write an RFC for that though, as it'll be a big new thing (ramifications for caching, session management, etc).

I'm not sure how making the live stack work with live accounts would work though, unless we maybe do something like add an OAuth app (to the production accounts system) for dev.gov.uk.

---

> If my team likes `app-account`

I think I've got a slight preference towards a separate stack for just accounts.  It keeps this stuff isolated to our team for now, which is good while we're still mulling over topics like how secrets should work in dev.

The way we're mostly doing things at the moment is using the live stack + an accounts system running in govuk-docker, so it'd be good to keep that if possible - for example, collections won't render /transition without a content item.  Needing to replicate data (or just edit the docker-compose file) to try out changes to that page would be kind of a pain.

---

> tech fortnightly

We've put a 45 minute slot in for the 1st of December tech fortnightly to have a bit more of an in-depth introduction to the accounts system and the tech it's built on, which will hopefully be useful to people.

Reply:
I think for now, we've avoided this one becuase we're looking at the checker that stores all it's own data... but there's interest in what wider personalisation would look like...

... which now i think about it will likely require a lot of content flying around when we're sense checking our development...
Good to have this on the radar early. Thanks!

Reply:
> The way we're mostly doing things at the moment is using the live stack + an accounts system running in govuk-docker, so it'd be good to keep that if possible - for example, collections won't render /transition without a content item. Needing to replicate data (or just edit the docker-compose file) to try out changes to that page would be kind of a pain. Or maybe we can live without that if it'll be confusing...

Yeah `/transition` is our first challenge here. I have an instinctive aversion to the idea of chaining these... so if we avoid a solution to this being anything like collections-app-account-live`.

I'm tempted to say shrug a bit about what the answer is for `collections` at the moment as the impact is small and we can test it in other ways, but if we start seeing greater needs to adapt that page then we should consider a dev setup early.

I suspect we'll still be working with a single content item for `/transition` for the forseeable, so perhaps we could find a way to just replicate that item locally with an `app-account` stack for any testing there?